flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java

```diff
@@ -0,0 +1,263 @@
+    public void triggerMaterialization() {
+        mailboxExecutor.execute(
+                () -> {
+                    // Only one materialization ongoing at a time
+                    if (!materializationOnGoing.compareAndSet(false, true)) {
+                        return;
+                    }
+
+                    SequenceNumber upTo = stateChangelogWriter.lastAppendedSequenceNumber().next();
+
+                    if (upTo.equals(materializedState.lastMaterializedTo())) {
+                        return;
+                    }
+
+                    RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture =
+                            keyedStateBackend.snapshot(
+                                    materializedId++, // TO DISCUSS: seems this does not matter,
+                                    // based on
+                                    System.currentTimeMillis(),
+                                    streamFactory,
+                                    CHECKPOINT_OPTIONS);
+
+                    // TODO: add metadata to log FLINK-23170
+                    asyncOperationsThreadPool.execute(
+                            () -> asyncMaterialization(materializedRunnableFuture, upTo));
+                },
+                "materialization");
+    }
+
+    private void asyncMaterialization(
+            RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture,
+            SequenceNumber upTo) {
+
+        FileSystemSafetyNet.initializeSafetyNetForThread();
+        try {
+            FutureUtils.runIfNotDoneAndGet(materializedRunnableFuture);
+            mailboxExecutor.execute(
+                    () -> {
+                        materializedState =
+                                new MaterializedState(
+                                        getMaterializedResult(materializedRunnableFuture.get()),
+                                        new ArrayList<>(),
+                                        upTo);
+
+                        checkState(
+                                materializationOnGoing.compareAndSet(true, false),
+                                "expect to finish materialization successfully, "
+                                        + "flag materializationOnGoing should be true before finishing.");
+                        retries.set(0);
+                    },
+                    "update materializedSnapshot up to changelog sequence number: {}",
+                    upTo);
+        } catch (Exception e) {
+            int retryTime = retries.incrementAndGet();
+
+            LOG.info(
+                    "Asynchronous part of materialization could not be completed for the {} time.",
+                    retryTime,
+                    e);
+
+            // ToDO: Double check
+            //  how the async phase is related to materialized state tracking
+            //  refer to AsyncCheckpointRunnable#run#catch
+            //  cleanup logic in StateUtil#discardStateFuture
```

Comment:
This should be fine (for tracking we need a Tracker anyway, and I think it should not affect this part of code).